enhancements/local-storage/automatic-partition.md

```diff
@@ -0,0 +1,193 @@
+
+### Goals
+
+- Partition the selected devices based on the partition policy specified by the user.
+- If a LocalVolumeSet has partitioning policy, it will create partitions on all disks that match the device filter.
+
+### Non-Goals
+
+- If anything changes later such as the partition policy or the device is resized, LSO will not add or modify any partitions.
+- Policy for wiping device to clean the partitions
+- Creating PVs based on both raw devices and partitions in the same LocalVolumeSet is not supported.
+
+### Risks and Mitigations
+- This will be handled by [LocalVolumeSet](https://github.com/openshift/enhancements/blob/master/enhancements/local-storage/automatic-detection.md#risks-and-mitigations)
+
+## Proposal
+
+The `local-storage-operator` is already capable of detecting local disks from the nodes and provisioning PVs out of them, but it can't partition them.
+The idea is to introduce the feature of partitioning the local disks and using the partitions to provision the PVs.
+
+The administrator can specify if he wants to partition the disks in the LocalVolumeSet CR and then LSO, after detecting the disk, will partition them first before provisioning PVs on those partitions.
+
+### Workflow for Local PV creation via `LocalVolumeSet CR`
+- The admin decides they want to create partitions on some devices in the cluster
+- He creates a LocalVolumeSet CR that includes the partitioning policy
+- Once the `LocalVolumeSet` is created, LSO reconciles the LocalVolumeSet CR (the same as designed for full device),
+- The controller will:
+  - configure the local-static-provisioner to make a new StorageClass based on certain directories on the selected nodes.
+  - assign diskmaker daemonset to the selected nodes.
+  - there will be one daemonset for each `LocalVolumeSet` CR.
+- The diskmaker daemon will find devices that match the disovery policy and check if partitoningSpec is present
+  - if partitonSpec is present, it will partition the devices.
+  - as soon as a device is partitoned
+    - the diskmaker will symlink the child partitions of the device into the directory that the local-static-provisioner is watching
+    - The local-static-provisioner will provision PVs on those partitions
+- The localVolumeset won't accept changes after its created and has a partitioningSpec
+
+
+## Design Details for Auto partitioning of local devices
+
+Updated `LocalVolumeSet` API sceme for Auto partition:
+
+
+```go
+
+type LocalVolumeSetSpec struct {
+ ...
+  // PartitioningSpec is the partitioning policy according to which the selected devices will be partitioned.
+  PartitioningSpec PartitoiningSpec `json:"partitioningSpec"`
+}
+
+type PartitioningSpec struct {
+  // PartSize determines the exact size of each partition.
+  PartSize *resource.Quantity `json:"partSize"`
```

Comment:
I feel like we need to have a `Count` as well. It looks like a waste to use the entire device all the time.

Reply:
If you're going to partition the disk, seems like you'd want to partition it all at the same time since the unpartitioned space wouldn't be usable. There could certainly be other options to indicate how partitions should be created, but I want to be clear about the scenarios to keep it simple.

Reply:
> If you're going to partition the disk, seems like you'd want to partition it all at the same time since the unpartitioned space wouldn't be usable. There could certainly be other options to indicate how partitions should be created, but I want to be clear about the scenarios to keep it simple.

I didn't understand by partition it all at the same time.
Do you suggest that we should not keep count for the current scenario?

Reply:
I am in favour of count as it is also a simple parameter to partition the disks.

Reply:
I guess the rationale is why should we use the entire device? If we device is an NVMe it could certainly be used by other applications. I don't think we (can?) restrict disks just for Ceph.

So if we decide that Ceph needs 3 partitions of 40GiB (making this up) then the rest can still be used by other applications.

Reply:
> So if we decide that Ceph needs 3 partitions of 40GiB (making this up) then the rest can still be used by other applications.

I understand the point, but the scenario is not clear to me.
Is this in case of `count` or `size`?

Reply:
Both.

Reply:
Got it.
So in case of `size` the user wants 30G partitions from a 100G disk.
We create 3 partitions of 30G and also create a partition for the remaining 10G right?
we can have an option that the user can select to partition the remaining space if he requires.

Reply:
The result of the partitioning will be that we have a PV created on each partition. We don't know who the consumer is, it might be Ceph or it might be some other app, depending on who requests the volumes. So why wouldn't we partition the whole disk? If you're partitioning it for PVs, do we need to allow apps to use it for non-PVs? I didn't think so, but deciding not to partition the whole device can certainly be part of the policy.

Reply:
For the parameters, is this the suggested policy in this thread? 
- If only `size` is specified, create partitions for the whole disk all of that size. If there is space remaining, create a PV for the remaining amount.
- If both `count` and `size` are specified, only create that number of partitions and leave the remaining disk unpartitioned.

That seems reasonable and simple to me.

Reply:
> For the parameters, is this the suggested policy in this thread?
> 
>     * If only `size` is specified, create partitions for the whole disk all of that size. If there is space remaining, create a PV for the remaining amount.
> 
>     * If both `count` and `size` are specified, only create that number of partitions and leave the remaining disk unpartitioned.
> 
My understanding of `count` is to create `count` no of partitions i.e., `Size of the entire device`/`count`
for example we have a device of total capacity of 128G and the value of `count` is 4, then 4 partitions of 32G each will be created.
> 
> That seems reasonable and simple to me.

Reply:
Ah, I was stuck thinking about the size as being a required setting. What about this?
- If only `size` is specified, create partitions for the whole disk all of that size. If there is space remaining, create a PV for the remaining amount.
- If only `count` is specified, divide the disk up into that many partitions and calculate the size.
- If both `count` and `size` are specified, only create that number of partitions of the desired size and leave the remaining disk unpartitioned.

Reply:
Got it